Apache branch: fixes to build system

chivakker opened the pull request:
These patches remove the warnings when running ./autogen.sh && make && make install && make uninstall && make distclean

chivakker committed:
configure: remove configure warnings

libyami build system is producing some warnings when running configure

* removed old functions used on configure.ac and Makefile.am files everywhere
* isolated unittests better as configure Plo/Po files are created even when
unittests are not enabled
* when removing the INCLUDES from Makefile.am files it is causing several
includes within cpp/h files to fail

TEST=autogen.sh should not produce any warning
TEST=make install should work
TEST=all functionality should be the same
TEST=make clean should work
TEST=make distclean should not trigger another configure
TEST=and it should erase the files intended to be erased

Change-Id: If0c14891655680d7c2f4bcd7825a3a83983cb63f
Signed-off-by: Daniel Charles <daniel.charles@intel.com>

chivakker committed:
v4l2-Makefile: avoid recursive compilation

recursive compilation to include libyami_v4l2 was causing
make distclean to report errors with files already removed

TEST=autogen.sh w/o --enable-v4l2, make, make distclean
TEST=all functionality should be the same and distclean
TEST=should finish properly

Change-Id: I7e6bba59934021fcb5e3d842c21a645e8d4dd50c
Signed-off-by: Daniel Charles <daniel.charles@intel.com>

chivakker committed:
unittest: fix all unsigned vs. signed comparisons

gcc 5.3.1 breaks compilation due to all these unsigned vs.
signed comparisons.  On previous gcc versions these should
be non-fatal warnings

TEST="compile with --enable-gtest and gcc 5.3.1"

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

uartie commented:
FYI, I was able to reproduce the signed/unsigned errors on GCC 4.9.2, too.

Today, I realized my build environment had user defined CFLAGS/CXXFLAGS which caused the compiler flags in `configure.ac` to get overshadowed on my builds (see bug #382).  In turn, I've been compiling without `-Werror -Wall` for who knows how long :-/.  I apologize for the false information about it likely being GCC 5.3.x that increased strictness (although it does seem to catch more of these).  You may want to remove the GCC version blurb from the commit message.  And I guess I should go fix #382 sooner than later ;-)  Cheers.

seanvk commented:
So let's make it a point to focus on compatibility at this time only
with Ubuntu LTS.  Let's be consistent on the compilers we use and test
with.

While it is fine to experiement across a wide range, I feel that it is
getting out of hand.

Thanks,

Sean

On 06/08/2016 07:39 PM, U. Artie Eoff wrote:

> FYI, I was able to reproduce the signed/unsigned errors on GCC 4.9.2, too.
> 
> Today, I realized my build environment had user defined CFLAGS/CXXFLAGS
> which caused the compiler flags in |configure.ac| to get overshadowed on
> my builds (see bug #382 https://github.com/01org/libyami/issues/382).
> In turn, I've been compiling without |-Werror -Wall| for who knows how
> long :-/. I apologize for the false information about it likely being
> GCC 5.3.x that increased strictness (although it does seem to catch more
> of these). You may want to remove the GCC version blurb from the commit
> message. And I guess I should go fix #382
> https://github.com/01org/libyami/issues/382 sooner than later ;-) Cheers.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/01org/libyami/pull/472#issuecomment-224786606, or
> mute the thread
> https://github.com/notifications/unsubscribe/AA0codvnkpiGcBaI2Pdsa-Ts5Q_5eQyAks5qJ3zZgaJpZM4IxaXe.

chivakker commented:
It is not my intention to change focus on compatibility or modify a regular building environment with autotools.  With this clear and Ubuntu being the target OS, this patch is fixing a problem Ubuntu 16.04 faces today.  As mentioned in the commit message, the version is expressed to make people aware of which version was used to test and that no other older version will regress due to this patch.

seanvk commented:
Sounds good.  Let's have PRC verify when back in.
lgtm +1

uartie commented:
I'm getting the following compiler error:

```
vaapiencoder_hevc.cpp:21:23: fatal error: bitWriter.h: No such file or directory
#include "bitWriter.h"
^
compilation terminated.
vaapiencoder_jpeg.cpp:23:26: fatal error: scopedlogger.h: No such file or directory
#include "scopedlogger.h"
^
compilation terminated.
```

uartie commented on capi/Makefile.am:
```diff
@@ -1,4 +1,4 @@
-INCLUDES = -I$(top_srcdir) \
+AM_CPPFLAGS = -I$(top_srcdir) \
```
You can remove AM_CPPFLAGS definition since  `-I$(top_srcdir)` should be automatic.  And `-I($top_srcdir)/interface` is unnecessary since you fixed the include paths in the source.

uartie replied:
https://www.gnu.org/software/automake/manual/html_node/Program-Variables.html

i.e. "Automake already provides some -I options automatically, in a separate variable that is also passed to every compilation that invokes the C preprocessor. In particular it generates ‘-I.’, ‘-I$(srcdir)’, and a -I pointing to the directory holding config.h (if you’ve used AC_CONFIG_HEADERS). "

We use AC_CONFIG_HEADERS, and config.h is in $(top_srcdir), thus we satisfy the last rule.

chivakker replied:
Yup I'll remove it.

uartie commented on codecparsers/Makefile.am:
```diff
@@ -1,4 +1,4 @@
-INCLUDES = -I$(top_srcdir)
+AM_CPPFLAGS = -I$(top_srcdir)
```
You can remove AM_CPPFLAGS since `-I$(top_srcdir)` should be automatic.

chivakker replied:
yes, to be removed

uartie commented on common/Makefile.am:
```diff
@@ -1,4 +1,4 @@
-INCLUDES = -I$(top_srcdir)
+AM_CPPFLAGS = -I$(top_srcdir)
```
You can remove AM_CPPFLAGS since `-I$(top_srcdir)` should be automatic.

chivakker replied:
yes, to be removed.

uartie commented on vaapi/Makefile.am:
```diff
@@ -1,4 +1,4 @@
-INCLUDES = -I$(top_srcdir)
+AM_CPPFLAGS = -I$(top_srcdir)
```
You can remove AM_CPPFLAGS since `-I$(top_srcdir)` should be automatic.

chivakker replied:
to be removed

chivakker commented:
I'll also fix the compilation issues, thanks for helping to properly test this.

chivakker commented:
So, I fixed the compilation issues, the comments from @uartie  and squashed the fixes into the original patch.  It needs more testing from different ends.  I'd wait for others to comment in as per @seanvk suggestion.

uartie commented:
lgtm

jayesht commented:
lgtm, validated this pull request on SKL ubuntu.

xuguangxin commented on codecparsers/bitReader_unittest.cpp:
```diff
@@ -48,12 +48,12 @@ BITREADER_TEST(NoOverflow)
     const uint32_t nBytes = data.size();
 
-    ASSERT_EQ(0, nBytes % (1024 * 1024 * 512));
+    ASSERT_EQ(0u, nBytes % (1024 * 1024 * 512));
```
It's very incontinence to add 'u' prefix for all  unsigned value. And it's hard to maintain.It's very likely we have someone do not use gcc 5.3.1 and did not pay attention to the singed unsigned comparison. He will introduce some unit test which break build for gcc 5.3.1.
Is it ok to ignore signed/unsigned compare warning in unit test?

uartie replied:
I don't think integer-suffixes are inconvenient.  Plus, they advertise the intentions of the test comparison better.  FYI, these warnings are happening on older GCC versions too.  IMO we shouldn't be ignoring any warnings.

xuguangxin replied:
Hi uarite:
I did not see this warning in my machine. Do I need turn on some compile flag? Could we add some compile flag to treat this as error in older gcc? I do not want some one add  a unittest in old gcc without error, then break other one who use new gcc.

uartie replied:
What version of GCC?  When you compile, do you see `-Wall -Werror` compiler flags in the compiler command line output?  If not, are you're setting C/CXXFLAGS in your environment? Environment settings will override the project settings (see #382 bug for more details on that).  So either add `-Wall -Werror` to your env C/CXXFLAGS or unset them.  You can also try adding `-Wsign-compare`.

Otherwise, older GCC (< 4.8.2 ?) might not be strict or smart enough to catch these with `-Wall -Werror` and/or `-Wsign-compare`.  If we are diligently validating and testing patches on all Ubuntu LTS releases (precise, trusty, and xenial - i.e. the "compatible" platforms) before accepting them (as we ought to), then breaking others shouldn't be too problematic.  If a patch gets in that breaks a newer GCC compile, then they are easy to fix.  These things just come with the territory of tools and software evolution.  I don't think it's critical to hold up this patchset for this concern.

xuguangxin replied:
here is my gcc version `gcc (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4`
and I do see`-Wall` and `-Werror` in Makefile.

I will try add some signed and unsigned compare in my h264 decoder

```
CC = gcc
CCDEPMODE = depmode=gcc3
CFLAGS = -g -O2 -Wno-unused-function  -Wno-cpp -Wall -Werror
CPP = gcc -E
CPPFLAGS =
CXX = g++
CXXCPP = g++ -E
CXXDEPMODE = depmode=gcc3
CXXFLAGS = -g -O2 -Wno-unused-function -Wno-cpp -Wall -Werror
CYGPATH_W = echo
DEFS = -DHAVE_CONFIG_H
DEPDIR = .deps
DLLTOOL = false

```

xuguangxin replied:
kind of wired, if i apply this patch

```
diff --git a/decoder/vaapidecoder_h264.cpp b/decoder/vaapidecoder_h264.cpp
index 646f500..66c8b70 100644
--- a/decoder/vaapidecoder_h264.cpp
+++ b/decoder/vaapidecoder_h264.cpp
@@ -1501,7 +1501,14 @@ YamiStatus VaapiDecoderH264::decodeCurrent()
     YamiStatus status = YAMI_SUCCESS;
     if (!m_currPic)
         return status;
-
+    int t = -1;
+    unsigned int s = 0;
+    if (t == 0u) {
+        printf("aaaaaa");
+    }
+    if (t == s) {
+        printf("failed");
+    }
     if (!m_currPic->decode()) {
         ERROR("decode %d failed", m_currPic->m_poc);
         // ignore it to let application continue to decode the next frame

```

gcc only report error for 
`if (t == s) {`
not for
`if (t == 0u) {`

```
make[2]: Entering directory `/home/thomas/projects/yami/libyami/decoder'
  CXX      libyami_decoder_la-vaapidecoder_h264.lo
vaapidecoder_h264.cpp: In member function 'YamiStatus YamiMediaCodec::VaapiDecoderH264::decodeCurrent()':
vaapidecoder_h264.cpp:1509:14: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     if (t == s) {
              ^
cc1plus: all warnings being treated as errors
make[2]: *** [libyami_decoder_la-vaapidecoder_h264.lo] Error 1
```

uartie replied:
`s` is not a constant, but 0u is.  If you change `s` type to `const unsigned int` then the warning goes away.

xuguangxin replied:
for 

> EXPECT_EQ(0u, reader.getPos());

reader.getPos is not const, right?

jayesht commented:
Ubuntu 15.10 with 4.2.0-36-generic #41-Ubuntu SMP Mon Apr 18 15:49:10 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux kernel also faces issue with unsigned comparison with GCC.

seanvk commented:
Jayesh your testing is out of date.  You should be using Ubuntu 16.04.  Please upgrade to latest LTS and re-verify.

seanvk commented:
-2 until tested

jayesht commented:
upgrading my system. Will post once verified with latest LTS.

jayesht commented:
Skylake machine:

$uname -a
4.4.0-24-generic #43-Ubuntu SMP Wed Jun 8 19:27:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux 

Stock GCC version 5.3.1 - verified that the patch fixes the unsigned comparison issue. 

LGTM.

xuguangxin commented:
merged as 5680ce21f5018d545ff82b36ba092a53ccbd36be. Since it will not break old gcc
